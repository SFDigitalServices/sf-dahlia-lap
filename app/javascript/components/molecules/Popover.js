import React, { useEffect, useState, useRef } from 'react'
import { usePopper } from 'react-popper'

// Source: https://www.30secondsofcode.org/react/s/use-click-outside
const useClickOutside = (ref, callback) => {
  const handleClick = (e) => {
    if (ref.current && !ref.current.contains(e.target)) {
      callback()
    }
  }

  useEffect(() => {
    document.addEventListener('click', handleClick)
    return () => {
      document.removeEventListener('click', handleClick)
    }
  })
}

const Popover = ({ buttonElement, children }) => {
  const [showPopper, setShowPopper] = useState(false)

  const buttonRef = useRef(null)
  const popperRef = useRef(null)

  useClickOutside(popperRef, () => setShowPopper(false))
  // the ref for the arrow must be a callback ref
  const [arrowRef, setArrowRef] = useState(null)

  const { styles, attributes } = usePopper(buttonRef.current, popperRef.current, {
    placement: 'bottom-end',
    modifiers: [
      {
        name: 'arrow',
        options: {
          element: arrowRef
        }
      },
      {
        name: 'offset',
        options: {
          offset: [0, 32]
        }
      }
    ]
  })

  return (
    <>
      {buttonElement({ ref: buttonRef, onClick: () => setShowPopper(!showPopper) })}
      {showPopper && (
        <div
          className='popper-container'
          ref={popperRef}
          style={styles.popper}
          {...attributes.popper}
        >
          <div ref={setArrowRef} id='arrow' />
          {children}
        </div>
      )}
    </>
  )
}

export default Popover
