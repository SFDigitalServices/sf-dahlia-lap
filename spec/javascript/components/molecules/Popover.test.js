import React from 'react'
import { shallow, mount } from 'enzyme'
import { act } from 'react-dom/test-utils'

import Popover from 'components/molecules/Popover'

const sampleButtonElement = ({ onClick, ref }) => (
  <button ref={ref} onClick={onClick}>
    Sample popover button
  </button>
)

const getWrapper = () =>
  shallow(
    <Popover buttonElement={sampleButtonElement}>
      <p>popover content</p>
    </Popover>
  )

describe('Popover', () => {
  it('should be closed by default', () => {
    const wrapper = getWrapper()
    expect(wrapper.find('p').exists()).toBeFalsy()
  })
  it('should renders the provided button as expected', () => {
    const wrapper = getWrapper()
    expect(wrapper).toMatchSnapshot()
  })

  it('should renders the tooltip content as expected', () => {
    const wrapper = getWrapper()
    wrapper.find('button').simulate('click')
    expect(wrapper).toMatchSnapshot()
  })
  it('should open and close the tooltip when button is clicked', () => {
    const wrapper = getWrapper()
    wrapper.find('button').simulate('click')
    expect(wrapper.find('p').exists()).toBeTruthy()
    wrapper.find('button').simulate('click')
    expect(wrapper.find('p').exists()).toBeFalsy()
  })

  describe('useClickOutside effect', () => {
    const map = {}
    document.addEventListener = jest.fn().mockImplementation((event, cb) => {
      map[event] = cb
    })

    const getWrapperWithOutsideElement = () =>
      mount(
        <div>
          <Popover buttonElement={sampleButtonElement}>
            <p>popover content</p>
          </Popover>
          <a id='something_else' />
        </div>
      )
    it('should close the tooltip with outside elements are clicked', () => {
      const wrapper = getWrapperWithOutsideElement()
      // Open the popover
      wrapper.find('button').simulate('click')
      expect(wrapper.find('p').exists()).toBeTruthy()
      // Click on something outside of the popover
      act(() => {
        map.click({ target: wrapper.find('#something_else').getDOMNode() })
      })
      wrapper.update()
      expect(wrapper.find('p').exists()).toBeFalsy()
    })

    it('should not close the tooltip when you click inside the tooltip', () => {
      const wrapper = getWrapperWithOutsideElement()
      // Open the popover
      wrapper.find('button').simulate('click')
      expect(wrapper.find('p').exists()).toBeTruthy()
      // Click on something inside of the popover
      act(() => {
        map.click({ target: wrapper.find('p').getDOMNode() })
      })
      wrapper.update()
      expect(wrapper.find('p').exists()).toBeTruthy()
    })
  })
})
