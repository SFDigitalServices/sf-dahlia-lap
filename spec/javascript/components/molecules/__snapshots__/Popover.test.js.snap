// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Popover should renders the provided button as expected 1`] = `
<Fragment>
  <button
    onClick={[Function]}
  >
    Sample popover button
  </button>
</Fragment>
`;

exports[`Popover should renders the tooltip content as expected 1`] = `
<Fragment>
  <button
    onClick={[Function]}
  >
    Sample popover button
  </button>
  <div
    className="popper-container"
    style={
      Object {
        "left": "0",
        "position": "absolute",
        "top": "0",
      }
    }
  >
    <div
      id="arrow"
    />
    <p>
      popover content
    </p>
  </div>
</Fragment>
`;
